Search Domains and DNS Resolver on Windows

flx5 opened the pull request:
Motivation:
On Windows with JDK 9+ the search domain and dns servers are missing.

Modification:

This PR uses JNI to fetch the DNS Servers from the operating system.

Result:

Fixes #12712 #11885.

hyperxpro commented:
More JNI 😐

normanmaurer commented on resolver-dns-native-windows/pom.xml:
```diff
@@ -0,0 +1,148 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2019 The Netty Project
```
```suggestion
  ~ Copyright 2022 The Netty Project
```

normanmaurer commented on resolver-dns-native-windows/pom.xml:
```diff
@@ -0,0 +1,148 @@
+  <parent>
+    <groupId>io.netty</groupId>
+    <artifactId>netty-parent</artifactId>
+    <version>4.1.85.Final-SNAPSHOT</version>
```
```suggestion
    <version>4.1.86.Final-SNAPSHOT</version>
```

normanmaurer commented on resolver-dns-native-windows/pom.xml:
```diff
@@ -0,0 +1,148 @@
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>com.googlecode.cmake-maven-project</groupId>
```
Can we use the hawtjini plugin so we are consistent with the rest of our native stuff ?

flx5 replied:
Will look into it, but that plugin had some trouble with newer Visual Studio Editions.

flx5 replied:
Replaced cmake with hawtjni

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/Connection.cpp:
```diff
@@ -0,0 +1,43 @@
+#include "connection.h"
```
missing license header

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/Connection.cpp:
```diff
@@ -0,0 +1,43 @@
+#pragma comment(lib, "ws2_32.lib")
+
+Connection::Connection(LPSOCKADDR sock_address) {
+	switch (sock_address->sa_family) {
```
we use 4 spaces... Please update

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/JavaInterface.cpp:
```diff
@@ -0,0 +1,63 @@
+#include "JavaNetworkAdapterClass.h"
+#include <stdexcept>
+
+#define NETTY_JNI_UTIL_JNI_VERSION JNI_VERSION_1_6
```
Can we just use c ? We dont use any c++ atm and I would like to keep it like this to make things easier for the maintainers .

flx5 replied:
Ported the native code to c.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/WindowsResolverDnsServerAddressStreamProvider.java:
```diff
@@ -0,0 +1,108 @@
+import io.netty.util.internal.StringUtil;
+
+import java.net.InetSocketAddress;
+import java.util.*;
```
we dont use wildcard imports

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+		return NULL;
+	}
+
+    return (*env)->NewObject(env, inetSocketAddressClass, inetSocketAddressCtor, host, port);
```
fix formatting

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+        return JNI_ERR;
+    }
+
+    if (loadClass(env, "java/util/LinkedList", &linkedListClass) != JNI_OK) {
```
Use an `ArrayList`... `LinkedList` shouldn't be used 99% of times.

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+    return result;
+}
+
+jobject searchDomainsToJava(JNIEnv* env, IP_ADAPTER_ADDRESSES* adapter) {
```
Same comment as above. Consider just returning Object[].

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.h:
```diff
@@ -0,0 +1,35 @@
+
+#ifndef _Included_io_netty_resolver_dns_windows_NativeWrapper
+#define _Included_io_netty_resolver_dns_windows_NativeWrapper
+#ifdef __cplusplus
```
as we dont use c++ anymore I think this cane be removed.

flx5 replied:
Removed.

normanmaurer commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+
+    free(inetSocketAddressCtor);
+    free(inetSocketAddressClass);
+}
```
missing EOL

normanmaurer commented on pom.xml:
```diff
@@ -1824,8 +1825,8 @@
-          <artifactId>maven-hawtjni-plugin</artifactId>
-          <version>1.14</version>
+          <artifactId>hawtjni-maven-plugin</artifactId>
+          <version>1.18</version>
```
Why do we need this change ? I think this is not really related and should be done as a separate PR.

flx5 replied:
You are right, this should be a separate PR.
The issue here is that the plugin requires additional configuration to support newer MSVC versions: https://github.com/fusesource/hawtjni/issues/76.

That configuration property was however only available with the newer hawtjni version, which conveniently was also renamed...

Old versions of visual studio don't seem to be available anymore from microsoft, so I could not compile and test this otherwise.

chrisvest commented on resolver-dns-native-windows/src/main/native-package/vs2010.custom.props:
```diff
@@ -0,0 +1,8 @@
+          <TreatWarningAsError>true</TreatWarningAsError>
+      </ClCompile>
+  </ItemDefinitionGroup>
+</Project>
```
Do we need this file? Also, I'm apprehensive about treating warnings as errors, since it can break the build down the line as newer compilers add more warnings.

flx5 replied:
No, this is not required. I don't typically write C, so I wanted to be sure that I don't introduce new issues. Thus I decided to treat warnings as errors, because otherwise hawtjni tends to hide them.

normanmaurer commented:
@flx5 please ping us when the comments were addressed

flx5 commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+        return JNI_ERR;
+    }
+
+    if (loadClass(env, "io/netty/resolver/dns/windows/NativeException", &nativeExceptionClass) != JNI_OK) {
```
Classes are now loaded with NETTY_JNI_UTIL_LOAD_CLASS.

flx5 commented on resolver-dns-native-windows/src/main/c/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+﻿cmake_minimum_required (VERSION 3.8)
```
Sorry, forgot about this one. Removed.

flx5 commented on resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NetworkAdapter.java:
```diff
@@ -0,0 +1,37 @@
+import java.net.InetSocketAddress;
+import java.util.List;
+
+public final class NetworkAdapter {
```
Yes, it can be. Marked as package private.

flx5 commented on resolver-dns-native-windows/src/main/java/io/netty/resolver/dns/windows/NativeException.java:
```diff
@@ -0,0 +1,22 @@
+ */
+package io.netty.resolver.dns.windows;
+
+public class NativeException extends RuntimeException {
```
Replaced with generic runtime exception. Or is there a better existing class?

flx5 commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+        pAddresses = NULL;
+    }
+
+    if (dwRetVal == NO_ERROR) {
+        *target = pAddresses;
+        return JNI_OK;
+    }
+
+    if (pAddresses != NULL) {
+        free(pAddresses);
+    }
```
That won't work. In the successful case the caller has to free the IP_ADAPTER_ADDRESSES structure. Otherwise the target Variable would point to already freed memory.

flx5 commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,380 @@
+    return (*env)->NewObject(env, inetSocketAddressClass, inetSocketAddressCtor, host, port);
+}
+
+jobject dnsServersToJava(JNIEnv* env, IP_ADAPTER_ADDRESSES* adapter) {
```
I tried using arrays at first, but then I'd have to iterate the c structure twice. One time to determine the size and another time to do the actual assignments. Not sure which variant is better here.

flx5 commented:
@normanmaurer @chrisvest Sorry, didn't get around to this for quite some time. Although we are using this project (and other open source projects) at work, I am not allowed to work on open source during work hours. 

So writing this PR on my own time which can therefore be a little bit slow going.

- Do you think it would make more sense to first introduce the hawtjni in another PR before this PR?
- Do you think it has a relevant performance / readability impact to use arrays instead of lists in the native code?

normanmaurer commented:
> @normanmaurer @chrisvest Sorry, didn't get around to this for quite some time. Although we are using this project (and other open source projects) at work, I am not allowed to work on open source during work hours.
> 
> So writing this PR on my own time which can therefore be a little bit slow going.

@flx5 no worries... 

> 
> * Do you think it would make more sense to first introduce the hawtjni in another PR before this PR?

Yes please.....

> * Do you think it has a relevant performance / readability impact to use arrays instead of lists in the native code?

We usually prefer to return arrays.

flx5 commented:
Hi @normanmaurer,

as discussed the hawtjni update has been placed in #13128.

I've also replaced the linked list with arrays and fixed some build issues.

Could you please review these changes?

Thank you!

flx5 commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,478 @@
+    unsigned short port;
+    jstring host;
+    
+	switch (sockAddr->sa_family) {
+	case AF_INET: {
+		char ip[INET_ADDRSTRLEN];
+		struct sockaddr_in* addr_in = (struct sockaddr_in*)sockAddr;
+
+        if (inet_ntop(AF_INET, &addr_in->sin_addr, ip, sizeof(ip)) == NULL) {
+            throwNativeException(env, "Could not convert Ipv4 to String");
+            return NULL;
+        }
+
+		port = ntohs(addr_in->sin_port);
+		host = (*env)->NewStringUTF(env, ip);
+		break;
+	}
+	case AF_INET6: {
+		char ip[INET6_ADDRSTRLEN];
+		struct sockaddr_in6* addr_in = (struct sockaddr_in6*)sockAddr;
+		
+        if (inet_ntop(AF_INET6, &addr_in->sin6_addr, ip, sizeof(ip)) == NULL) {
+            throwNativeException(env, "Could not convert Ipv6 to String");
+            return NULL;
+        }
+
+		port = ntohs(addr_in->sin6_port);
+		host = (*env)->NewStringUTF(env, ip);
+		break;
+	}
+	default:
+		throwNativeException(env, "Unknown address family %d", sockAddr->sa_family);
+		return NULL;
+	}
```
Oops. Sorry wrong IDE configuration re: tabs vs spaces

flx5 commented on resolver-dns-native-windows/src/main/c/src/netty_window_dns_native.c:
```diff
@@ -0,0 +1,478 @@
+#pragma comment(lib, "IPHLPAPI.lib")
+#pragma comment(lib, "ws2_32.lib")
+
+#define NETTY_JNI_UTIL_JNI_VERSION JNI_VERSION_1_6
```
Yes, that is from netty_jni_util.h
That was a remainder from the start when I didn't know about the util lib.
Removed that here.

normanmaurer commented:
@flx5 seems like you didn't sign our ICLA yet ? https://netty.io/s/icla ?

flx5 commented:
@normanmaurer Sorry, did not see that. I've signed it now.

normanmaurer commented:
@flx5 thanks so much ... this is amazing
